k-distribution/k-tutorial/1_basic/03_parsing/README.md

```diff
@@ -1 +1,362 @@
+```k
+module LESSON-03-C
+  syntax Color ::= "Yellow" "(" ")" | "Blue" "(" ")"
+  syntax Fruit ::= "Banana" "(" ")" | "Blueberrry" "(" ")"
+  syntax Color ::= "colorOf" "(" Fruit ")" [function]
+endmodule
+```
+
+In this example, the sorts of the argument to the function are unchanged, but
+everything else has been wrapped in double quotation marks. This is because
+in BNF notation, we distinguish between two types of **production items**:
+**terminals** and **non-terminals**. A terminal represents simply a literal
+string of characters that is verbatim part of the syntax of that production.
+A non-terminal, conversely, represents a sort name, where the syntax of that
+production accepts any valid term of that sort at that position.
+
+This is why, when we wrote the program `colorOf(Banana())`, `krun` was able to
+execute that program: because it represented a term of sort `Color` that was
+parsed and interpreted by K's interpreter. In other words, `krun` parses and
+interprets terms according to the grammar defined by the developer. It is
+automatically converted into an AST of that term, and then the `colorOf`
+function is evaluated using the function rules provided in the definition.
+
+Bringing us back to the file `lesson-03-a.k`, we can see that this grammar
+has given a simple BNF grammar for expressions over booleans. We have defined
+constructors corresponding to the boolean values true and false, and functions
+corresponding to the boolean operators for and, or, not, and xor. We have also
+given a syntax for each of these functions based on their syntax in the `C`
+programming language. As such, we can now write programs in the simple language
+we have defined.
+
+Input the following program into your editor as `and.bool` in the same
+directory:
+
+```
+true && false
+```
+
+We cannot interpret this program yet, because we have not given rules defining
+the meaning of the `&&` function yet, but we can parse it. To do this, you can
+run (from the same  directory):
+
+```
+kast --output kore and.bool
+```
+
+`kast` is K's just-in-time parser. It will generate a grammar from your K
+definition on the fly and use it to parse the program passed on the command
+line. The `--output` flag controls how the resulting AST is represented; don't
+worry about the possible values yet, just use `kore`.
+
+You ought to get the following AST printed on standard output, minus the
+formatting:
+
+```
+inj{SortBoolean{}, SortKItem{}}(
+  Lbl'UndsAnd-And-UndsUnds'LESSON-03-A'Unds'Boolean'Unds'Boolean'Unds'Boolean{}(
+    Lbltrue'Unds'LESSON-03-A'Unds'Boolean{}(),
+    Lblfalse'Unds'LESSON-03-A'Unds'Boolean{}()
+  )
+)
+```
+
+Don't worry about what exactly this means yet, just understand that it
+represents the AST of the program that you just parsed. You ought to be able
+to recognize the basic shape of it by seeing the words `true`, `false`, and
+`And` in there. This is **Kore**, the intermediate representation of K, and we
+will cover it in detail later.
```

Comment:
I think you should have them do `--output kast` as well, and have them compare the Kore and the KAST, make sure they can map the `_&&__LESSON-03-A_Boolean_Boolean_Boolean` symbol to the munged version. You could just say "Now try `--output kast` and see how they compare" here as something they can try.

Reply:
Just giving them something to do like this will force them to actually read the output.